tmol/io/pose_stack_deconstruction.py

```diff
@@ -0,0 +1,245 @@
+
+    expanded_coords, _ = pose_stack.expand_coords()
+    block_atom_is_real = torch.zeros(
+        (n_poses, max_n_res, max_n_atoms_per_res), dtype=torch.bool, device=device
+    )
+    block_atom_is_real[is_real_block] = pbt.atom_is_real[real_bt_inds64]
+    (
+        nz_pose_ind_for_real_atom,
+        nz_res_ind_for_real_atom,
+        nz_block_atom_ind_for_real_atom,
+    ) = torch.nonzero(block_atom_is_real, as_tuple=True)
+
+    canonical_atom_inds_for_bt = pbt.canonical_atom_ind_map[real_bt_inds64]
+    is_real_canonical_atom_ind_for_bt = canonical_atom_inds_for_bt != -1
+    real_canonical_atom_inds_for_bt = canonical_atom_inds_for_bt[
+        is_real_canonical_atom_ind_for_bt
+    ]
+    cf_coords = torch.full(
+        (n_poses, max_n_res, max_n_canonical_atoms, 3),
+        torch.nan,
+        dtype=torch.float32,
+        device=device,
+    )
+
+    cf_coords[
+        nz_pose_ind_for_real_atom,
+        nz_res_ind_for_real_atom,
+        real_canonical_atom_inds_for_bt,
+    ] = expanded_coords[block_atom_is_real]
+    atom_is_present = torch.zeros(
+        (n_poses, max_n_res, max_n_canonical_atoms), dtype=torch.bool, device=device
+    )
+    atom_is_present[
+        nz_pose_ind_for_real_atom,
+        nz_res_ind_for_real_atom,
+        real_canonical_atom_inds_for_bt,
+    ] = True
+
+    disulfide_conn_ind = torch.full(
+        (n_poses, max_n_res), -1, dtype=torch.int64, device=device
+    )
+    disulfide_conn_ind[is_real_block] = pbt.canonical_dslf_conn_ind[real_bt_inds64]
+    is_real_disulfide_conn = disulfide_conn_ind != -1
+    nz_pose_ind_for_dslf, nz_res_ind_for_dslf = torch.nonzero(
+        is_real_disulfide_conn, as_tuple=True
+    )
+    dslf_partner = pose_stack.inter_residue_connections[
+        nz_pose_ind_for_dslf,
+        nz_res_ind_for_dslf,
+        disulfide_conn_ind[is_real_disulfide_conn],
+        0,
+    ]
+
+    def _u1(x):
+        return x.unsqueeze(1)
+
+    redundant_dslf_tuples = torch.cat(
+        (_u1(nz_pose_ind_for_dslf), _u1(nz_res_ind_for_dslf), _u1(dslf_partner)), dim=1
+    )
+    is_non_redundant_dslf_tuple = (
+        redundant_dslf_tuples[:, 1] < redundant_dslf_tuples[:, 2]
+    )
+    disulfides = redundant_dslf_tuples[is_non_redundant_dslf_tuple, :]
+
+    if chain_id is None:
+        chain_id = torch.tensor(
+            chain_inds_for_pose_stack(pose_stack), dtype=torch.int32, device=device
+        )
+    else:
+        assert chain_id.shape == (n_poses, max_n_res)
+
+    res_not_connected = determine_res_not_connected_from_pose_stack(
+        pose_stack, chain_id, is_real_block, real_bt_inds64
+    )
+
+    return (
+        chain_id,
+        cf_res_types,
+        cf_coords,
+        atom_is_present,
+        disulfides,
+        res_not_connected,
+    )
+
+
+def _annotate_packed_block_types_w_termini_types(pbt: PackedBlockTypes):
+    # TEMP!
+    # for now: a bt is a down/up terminus if it includes
```

Comment:
how is this going to be changed in the future? (why is it TEMP)?

Reply:
The code I'm working on to generalize the termini patches looks at all the Patches in the PatchedChemicalDatabase and notes which patches remove the "{<down>}" or "{<up>}" atoms. It can then determine if a given patched block type is a termini variant by seeing what variant-display-names the variant's name includes. Right now, instead of using that code, I have effectively hard coded "there are only two types of patches that change the termini of residues and they are the "nterm" and "cterm" patches."